src/main/scala/org/graphframes/Pregel.scala

```diff
@@ -0,0 +1,236 @@
+   *                `Pregel.src("src_col_name)`, `Pregel.edge("edge_col_name)`,
+   *                `Pregel.dst("dst_col_name)`
+   */
+  def sendMsgToSrc(msgExpr: Column): this.type = {
+    sendMsgs += Tuple2(Pregel.src("id"), msgExpr)
+    this
+  }
+
+
+  /**
+   * Set the message column. In each round of pregel, each triplet
+   * (src-edge-dst) will generate zero or one message and the message will
+   * be sent to the dst vertex of this triplet.
+   *
+   * @param msgExpr The message expression. It is a sql expression and it
+   *                can reference all propertis in the triplet, in the way
+   *                `Pregel.src("src_col_name)`, `Pregel.edge("edge_col_name)`,
+   *                `Pregel.dst("dst_col_name)`
+   */
+  def sendMsgToDst(msgExpr: Column): this.type = {
+    sendMsgs += Tuple2(Pregel.dst("id"), msgExpr)
+    this
+  }
+
+  /**
+   * Set the aggregation expression which used to aggregate messages which received
+   * by each vertex.
+   *
+   * @param aggExpr The aggregation expression, such as `sum(Pregel.msgCol)`
+   */
+  def aggMsgs(aggExpr: Column): this.type = {
+    aggMsgsCol = aggExpr
+    this
+  }
+
+  /**
+   * After set a series of things via above methods, then call this method to run
+   * pregel, and it will return the result vertex dataframe, which will include all
+   * updated columns in the final rounds of pregel.
+   *
+   * @return the result vertex dataframe
+   */
+  def run(): DataFrame = {
+    val sendMsgsColList = sendMsgs.toList.map { case (id, msg) =>
+      struct(id.as("id"), msg.as("msg"))
+    }
+
+    val initVertexCols = withVertexColumnList.toList.map { case (colName, initExpr, _) =>
+      initExpr.as(colName)
+    }
+    val updateVertexCols = withVertexColumnList.toList.map { case (colName, _, updateExpr) =>
+      when(Pregel.msg.isNotNull, updateExpr).otherwise(col(colName)).as(colName)
+    }
+
+    var vertexUpdateColDF = graph.vertices.select((col("id") :: initVertexCols): _*)
+    var vertices = graph.vertices.join(vertexUpdateColDF, "id")
+    val edges = graph.edges
+
+    var iteration = 1
+
+    val shouldCheckpoint = checkpointInterval > 0
+
+    require(sendMsgs.length > 0, "We need to set at least one message expression for pregel running.")
+
+    while (iteration <= maxIter) {
+      val tripletsDF = vertices.select(struct(col("*")).as("src"))
+        .join(edges.select(struct(col("*")).as("edge")), Pregel.src("id") === Pregel.edge("src"))
+        .join(vertices.select(struct(col("*")).as("dst")), Pregel.edge("dst") === Pregel.dst("id"))
+
+      var msgDF: DataFrame = tripletsDF
+        .select(explode(array(sendMsgsColList: _*)).as("msg"))
+        .select(col("msg.id"), col("msg.msg").as(Pregel.MSG_COL_NAME))
+
+      val newValueDF = msgDF
+        .filter(Pregel.msg.isNotNull)
+        .groupBy("id")
+        .agg(aggMsgsCol.as(Pregel.MSG_COL_NAME))
+
+      val verticesWithMsg = vertices.join(newValueDF, Seq("id"), "left_outer")
```

Comment:
do not join old vertices here

Reply:
If this, I try this way:
```
aggMsgDF = ...
// do cache & checkpoint on `aggMsgDF`
val verticesWithMsg = vertices.join(aggMsgDF, Seq(ID), "left_outer")
vertices = verticesWithMsg.select((originalVertexCols ++ updateCols): _*)
```
But it bring issues that current iteration vertices DF depends on previous one and the dependency cannot be cut by the cache/persist.

So, 
 - if we only cache the aggregated msg DF in each iteration, then we can remove one join here, **but it bring issues that the DF dependency cannot be cut by the cache/checkpoint and cause computation lineage to be long.**
 - if we only want to cache the additional columns for vertex DF in each iteration, then we have to do the two joins (one before cache/persist, one after it)
 - if we cache the whole vertices (including original columns and additional columns), then we can remove one join here.

I try every way but I do not find the best solution which can satisfy:
 - only cache additional columns or only cache aggregated msg column
 - remove one join here
 - do not cause computation lineage increased in each iteration.

Could you help think of a better solution (with pseudo code) ?